File: python/2684.MaximumNumberofMovesinaGrid-2.py
from typing import List, Tuple


class Solution:
    def maxMoves(self, grid: List[List[int]]) -> int:
        """
        Finds the maximum number of moves in a grid.

        Args:
          grid (List[List[int]]): The input grid of integers.

        Returns:
          int: The maximum number of moves.
        """
        if not grid:
            return 0

        rows, cols = len(grid), len(grid[0])
        dp = [[-1] * cols for _ in range(rows)]

        def helper(ri: int, cj: int) -> int:
            """
            Helper function to calculate the maximum number of moves from a given cell.

            Args:
              ri (int): The row index of the current cell.
              cj (int): The column index of the current cell.

            Returns:
              int: The maximum number of moves from the current cell.
            """
            if dp[ri][cj] != -1:
                return dp[ri][cj]

            directions: List[Tuple[int, int]] = [
                (ri - 1, cj + 1),
                (ri, cj + 1),
                (ri + 1, cj + 1),
            ]
            max_moves = 0

            for dr, dc in directions:
                if 0 <= dr < rows and 0 <= dc < cols and grid[dr][dc] > grid[ri][cj]:
                    max_moves = max(max_moves, helper(dr, dc) + 1)

            dp[ri][cj] = max_moves
            return dp[ri][cj]

        return max(helper(i, 0) for i in range(rows))

File: typescript/2684.MaximumNumberofMovesinaGrid-2.ts
/**
 * Finds the maximum number of moves in a grid.
 *
 * @param {number[][]} grid - The input grid of integers.
 * @returns {number} - The maximum number of moves.
 */
function maxMoves(grid: number[][]): number {
  // If the grid is empty, return 0 as there are no moves possible.
  if (grid.length === 0) return 0;

  // Get the number of rows and columns in the grid.
  const 
    rows: number = grid.length,
    cols: number = grid[0].length,
    // Initialize a 2D array to store the maximum number of moves from each cell.
    dp: number[][] = Array.from(
      { length: rows },
      () => Array(cols).fill(-1)
    );
  
  /**
   * Helper function to calculate the maximum number of moves from a given cell.
   *
   * @param {number} ri - The row index of the current cell.
   * @param {number} cj - The column index of the current cell.
   * @returns {number} - The maximum number of moves from the current cell.
   */
  const helper = (ri: number, cj: number): number => {
    // If the value is already computed, return it.
    if (dp[ri][cj] !== -1) return dp[ri][cj];
    
    // Define the possible directions to move in the grid.
    const directions: number[][] = [
      [ri - 1, cj + 1], // Move to the top-right cell.
      [ri, cj + 1],     // Move to the right cell.
      [ri + 1, cj + 1], // Move to the bottom-right cell.
    ];
    let maxMoves: number = 0;
    
    // Iterate through each possible direction.
    for (const [dr, dc] of directions) {
      // Check if the new cell is within the grid bounds and has a greater value.
      if (
        dr >= 0 &&
        dr < rows &&
        dc >= 0 &&
        dc < cols &&
        grid[dr][dc] > grid[ri][cj]
      ) {
        // Recursively calculate the maximum moves from the new cell and update the maxMoves.
        maxMoves = Math.max(maxMoves, helper(dr, dc) + 1);
      }
    }
    // Store the computed value in the dp array.
    dp[ri][cj] = maxMoves;
    return dp[ri][cj];
  };

  let result = 0;
  // Calculate the maximum number of moves starting from each cell in the first column.
  for (let i = 0; i < rows; i++) {
    result = Math.max(result, helper(i, 0));
  }

  // Return the maximum number of moves found.
  return result;
};
